File: clients/typescript/ctxd-client/src/http.ts
/**
 * HTTP admin client for the ctxd daemon.
 *
 * Hand-written types matching `docs/api/openapi.yaml`. We deliberately
 * avoid OpenAPI codegen for v0.3 — the surface is small, the shapes
 * are stable, and a hand-rolled client is easier to read in an
 * incident.
 *
 * Auth model: every constructor accepts an optional bearer token.
 * When set, it is attached as `Authorization: Bearer <token>` on
 * every request. Endpoints documented as open (`/health`,
 * `/v1/grant`, `/v1/stats`, `/v1/approvals`) tolerate the header
 * anyway, so we send it unconditionally when configured.
 *
 * Transport: relies on the runtime's global `fetch` — Node 20+ ships
 * undici under the hood; modern browsers ship native fetch. Same code
 * runs in both.
 *
 * Logging: this module never logs the bearer token, capability bytes,
 * or signature material. Status codes and body excerpts are fair
 * game; the secret material is not.
 */
import { httpStatusError, UnexpectedWireResponseError } from "./errors.js";
import { type Operation } from "./operation.js";

/** `GET /health` response body. */
export interface HealthInfo {
  status: string;
  version: string;
}

/** `GET /v1/stats` response body. Future versions may add fields. */
export interface StatsInfo {
  subject_count: number;
  [k: string]: unknown;
}

/** `GET /v1/peers` response item. */
export interface PeerInfo {
  peer_id: string;
  url: string;
  public_key: string;
  subject_patterns: string[];
  added_at: string;
  last_seen_at?: string | null;
}

/** Per-call HTTP timeout. Long enough to forgive a slow loopback under contention,
 *  short enough that a misconfigured URL fails loudly rather than hanging. */
export const DEFAULT_TIMEOUT_MS = 10_000;

/** Construction options for the HTTP admin client. */
export interface HttpAdminClientOptions {
  /** Bearer token attached to every request. */
  token?: string;
  /** Per-call HTTP timeout in milliseconds. Default: {@link DEFAULT_TIMEOUT_MS}. */
  timeoutMs?: number;
  /**
   * Override the global `fetch` (e.g. for Bun / older runtimes /
   * test doubles). Defaults to the global `fetch`.
   */
  fetch?: typeof fetch;
}

/** Async HTTP admin client. */
export class HttpAdminClient {
  private readonly base: string;
  private token: string | undefined;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(baseUrl: string, options: HttpAdminClientOptions = {}) {
    this.base = baseUrl.replace(/\/+$/, "");
    this.token = options.token;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    const f = options.fetch ?? globalThis.fetch;
    if (typeof f !== "function") {
      throw new Error(
        "no global fetch found; pass `fetch` via HttpAdminClientOptions on this runtime",
      );
    }
    // Re-bind so `this` doesn't get lost on detached calls (matters
    // for some polyfills + browsers).
    this.fetchImpl = f.bind(globalThis);
  }

  /** Base URL the client is pointed at (with any trailing slash stripped). */
  get baseUrl(): string {
    return this.base;
  }

  /** Bearer token attached to every request, or `undefined`. */
  get bearerToken(): string | undefined {
    return this.token;
  }

  /**
   * Attach a capability token to all admin calls. Replaces any
   * previously attached token.
   */
  withToken(token: string): this {
    this.token = token;
    return this;
  }

  /** `GET /health` — daemon liveness + version probe. */
  async health(): Promise<HealthInfo> {
    const json = (await this.getJson("/health")) as Record<string, unknown>;
    return {
      status: String(json.status),
      version: String(json.version),
    };
  }

  /** `GET /v1/stats` — basic store statistics. */
  async stats(): Promise<StatsInfo> {
    const json = (await this.getJson("/v1/stats")) as Record<string, unknown>;
    const sc = json.subject_count;
    if (typeof sc !== "number") {
      throw new UnexpectedWireResponseError(
        `stats response missing numeric subject_count: ${JSON.stringify(json)}`,
      );
    }
    return { ...json, subject_count: sc };
  }

  /**
   * `POST /v1/grant` — mint a capability token.
   *
   * Returns the base64-encoded biscuit token. The optional `expiresAt`
   * is converted to `expires_in_secs` server-side; we clamp at 1 so
   * passing a past timestamp surfaces a clear 400 from the server
   * rather than an "expires_in_secs: 0" rejection.
   */
  async grant(
    subject: string,
    operations: ReadonlyArray<Operation | string>,
    expiresAt?: Date,
  ): Promise<string> {
    const ops = operations.map((o) => String(o));
    const body: Record<string, unknown> = {
      subject,
      operations: ops,
    };
    if (expiresAt instanceof Date) {
      const deltaSec = Math.floor((expiresAt.getTime() - Date.now()) / 1000);
      body.expires_in_secs = Math.max(1, deltaSec);
    }
    const resp = (await this.postJson("/v1/grant", body)) as Record<
      string,
      unknown
    >;
    const token = resp.token;
    if (typeof token !== "string") {
      throw new UnexpectedWireResponseError(
        `grant response missing string \`token\` field: ${JSON.stringify(resp)}`,
      );
    }
    return token;
  }

  /** `GET /v1/peers` — list federation peers (admin). */
  async peers(): Promise<PeerInfo[]> {
    const json = (await this.getJson("/v1/peers")) as Record<string, unknown>;
    const list = Array.isArray(json.peers) ? json.peers : [];
    return list.map((p: unknown) => parsePeer(p));
  }

  /** `DELETE /v1/peers/{peerId}` — remove a federation peer (admin). */
  async peerRemove(peerId: string): Promise<void> {
    await this.delete(`/v1/peers/${encodeURIComponent(peerId)}`);
  }

  // ----- Internals -----

  private headers(): Record<string, string> {
    const h: Record<string, string> = {
      Accept: "application/json",
    };
    if (this.token) {
      h.Authorization = `Bearer ${this.token}`;
    }
    return h;
  }

  private async getJson(path: string): Promise<unknown> {
    const url = `${this.base}${path}`;
    const resp = await this.fetchWithTimeout(url, {
      method: "GET",
      headers: this.headers(),
    });
    return decodeJson(resp);
  }

  private async postJson(
    path: string,
    body: Record<string, unknown>,
  ): Promise<unknown> {
    const url = `${this.base}${path}`;
    const resp = await this.fetchWithTimeout(url, {
      method: "POST",
      headers: {
        ...this.headers(),
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    return decodeJson(resp);
  }

  private async delete(path: string): Promise<void> {
    const url = `${this.base}${path}`;
    const resp = await this.fetchWithTimeout(url, {
      method: "DELETE",
      headers: this.headers(),
    });
    if (resp.status < 200 || resp.status >= 300) {
      const text = await safeText(resp);
      throw httpStatusError(resp.status, text);
    }
  }

  private async fetchWithTimeout(
    url: string,
    init: RequestInit,
  ): Promise<Response> {
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      return await this.fetchImpl(url, { ...init, signal: controller.signal });
    } finally {
      clearTimeout(t);
    }
  }
}

async function decodeJson(resp: Response): Promise<unknown> {
  if (resp.status >= 200 && resp.status < 300) {
    if (resp.status === 204) return null;
    const text = await resp.text();
    if (text.length === 0) return null;
    try {
      return JSON.parse(text);
    } catch (e) {
      throw new UnexpectedWireResponseError(
        `expected JSON body, got ${(e as Error).message}: ${text.slice(0, 200)}`,
      );
    }
  }
  const text = await safeText(resp);
  throw httpStatusError(resp.status, text);
}

async function safeText(resp: Response): Promise<string> {
  try {
    return await resp.text();
  } catch {
    return "";
  }
}

function parsePeer(p: unknown): PeerInfo {
  if (!p || typeof p !== "object") {
    throw new UnexpectedWireResponseError(
      `peer entry not an object: ${JSON.stringify(p)}`,
    );
  }
  const r = p as Record<string, unknown>;
  return {
    peer_id: String(r.peer_id ?? ""),
    url: String(r.url ?? ""),
    public_key: String(r.public_key ?? ""),
    subject_patterns: Array.isArray(r.subject_patterns)
      ? r.subject_patterns.map(String)
      : [],
    added_at: String(r.added_at ?? ""),
    last_seen_at:
      typeof r.last_seen_at === "string" ? r.last_seen_at : null,
  };
}

File: clients/rust/ctxd-client/src/http.rs
//! HTTP admin client for the ctxd daemon.
//!
//! Hand-written types matching `docs/api/openapi.yaml`. Codegen
//! (progenitor, openapi-generator, etc.) is intentionally avoided for
//! v0.3 — the surface is small, the shapes are stable, and a
//! hand-rolled client is easier to read in an incident.
//!
//! Auth model: every constructor accepts an optional bearer token.
//! When set, it is attached as `Authorization: Bearer <token>` on
//! every request. Endpoints documented as open (`/health`, `/v1/grant`,
//! `/v1/stats`, `/v1/approvals`) tolerate the header being present
//! anyway, so we send it unconditionally when configured.

use std::time::Duration;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::errors::CtxdError;

/// Default timeout for individual HTTP admin calls. Picked to be long
/// enough to forgive a slow loopback under contention but short enough
/// that a misconfigured URL fails the SDK loudly rather than hanging
/// the host application.
pub const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);

/// HTTP admin client.
#[derive(Debug, Clone)]
pub struct HttpAdminClient {
    base: String,
    inner: reqwest::Client,
    token: Option<String>,
}

impl HttpAdminClient {
    /// Construct a client pointed at the given base URL.
    ///
    /// The URL must include the scheme and host (e.g.
    /// `http://127.0.0.1:7777`). Trailing slashes are stripped so
    /// callers can pass either form.
    pub fn new(base_url: &str) -> Result<Self, CtxdError> {
        let inner = reqwest::Client::builder()
            .timeout(DEFAULT_TIMEOUT)
            .build()?;
        Ok(Self {
            base: base_url.trim_end_matches('/').to_string(),
            inner,
            token: None,
        })
    }

    /// Attach a capability token. Sent as `Authorization: Bearer
    /// <token>` on every subsequent call.
    pub fn with_token(mut self, token: impl Into<String>) -> Self {
        self.token = Some(token.into());
        self
    }

    /// Issue a GET against `<base>/<path>` and decode the JSON body.
    async fn get_json<T: for<'de> Deserialize<'de>>(&self, path: &str) -> Result<T, CtxdError> {
        let url = format!("{}{}", self.base, path);
        let mut req = self.inner.get(&url);
        if let Some(t) = &self.token {
            req = req.bearer_auth(t);
        }
        let resp = req.send().await?;
        let status = resp.status();
        if status.is_success() {
            let body = resp.json::<T>().await?;
            Ok(body)
        } else {
            let body = resp.text().await.unwrap_or_default();
            Err(CtxdError::http_status(status, body))
        }
    }

    /// Issue a POST with a JSON body and decode the JSON response.
    async fn post_json<B: Serialize, T: for<'de> Deserialize<'de>>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T, CtxdError> {
        let url = format!("{}{}", self.base, path);
        let mut req = self.inner.post(&url).json(body);
        if let Some(t) = &self.token {
            req = req.bearer_auth(t);
        }
        let resp = req.send().await?;
        let status = resp.status();
        if status.is_success() {
            let body = resp.json::<T>().await?;
            Ok(body)
        } else {
            let body = resp.text().await.unwrap_or_default();
            Err(CtxdError::http_status(status, body))
        }
    }

    /// Issue a DELETE; expects a 204 No Content on success.
    async fn delete_no_content(&self, path: &str) -> Result<(), CtxdError> {
        let url = format!("{}{}", self.base, path);
        let mut req = self.inner.delete(&url);
        if let Some(t) = &self.token {
            req = req.bearer_auth(t);
        }
        let resp = req.send().await?;
        let status = resp.status();
        if status.is_success() {
            Ok(())
        } else {
            let body = resp.text().await.unwrap_or_default();
            Err(CtxdError::http_status(status, body))
        }
    }

    /// `GET /health` — daemon liveness + version probe.
    pub async fn health(&self) -> Result<HealthInfo, CtxdError> {
        self.get_json("/health").await
    }

    /// `GET /v1/stats` — basic store statistics.
    pub async fn stats(&self) -> Result<StatsInfo, CtxdError> {
        self.get_json("/v1/stats").await
    }

    /// `POST /v1/grant` — mint a capability token.
    ///
    /// Returns the base64-encoded token (the `token` field of the
    /// daemon's response).
    pub async fn grant(
        &self,
        subject: &str,
        operations: &[Operation],
        expires_at: Option<DateTime<Utc>>,
    ) -> Result<String, CtxdError> {
        let expires_in_secs = expires_at.map(|t| {
            let now = Utc::now();
            let delta = t.signed_duration_since(now).num_seconds();
            // The daemon requires a positive integer; clamp at 1s if a
            // caller hands us a past timestamp so we surface a clear
            // 400 from the server rather than a confusing
            // "expires_in_secs: 0" rejection.
            delta.max(1)
        });
        let body = GrantRequest {
            subject: subject.to_string(),
            operations: operations.iter().map(|o| o.as_wire_str().to_string()).collect(),
            expires_in_secs,
        };
        let resp: GrantResponse = self.post_json("/v1/grant", &body).await?;
        Ok(resp.token)
    }

    /// Revoke a capability token by id.
    ///
    /// **Note:** the daemon does not yet expose a REST revoke endpoint
    /// in v0.3. The wire-protocol `Revoke` verb is the supported path.
    /// This HTTP-level method exists for shape parity with the wire
    /// client; today it returns
    /// [`CtxdError::HttpStatus`] with status `405` so callers can
    /// detect-and-fall-back. Tracked for v0.4.
    pub async fn revoke(&self, _token_id: &str) -> Result<(), CtxdError> {
        Err(CtxdError::HttpStatus {
            status: 405,
            body: "HTTP revoke is not implemented; use the wire protocol Revoke verb".to_string(),
        })
    }

    /// `GET /v1/peers` — list federation peers (admin).
    pub async fn peers(&self) -> Result<Vec<PeerInfo>, CtxdError> {
        let resp: PeerListResponse = self.get_json("/v1/peers").await?;
        Ok(resp.peers)
    }

    /// `DELETE /v1/peers/{peer_id}` — remove a federation peer (admin).
    pub async fn peer_remove(&self, peer_id: &str) -> Result<(), CtxdError> {
        let path = format!("/v1/peers/{peer_id}");
        self.delete_no_content(&path).await
    }
}

/// Operations a capability token can authorize.
///
/// Mirrors `ctxd_cap::Operation` and the OpenAPI spec's
/// `operations[]` enum. Wire serialization matches the daemon's
/// snake_case names exactly.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Operation {
    /// Read events under a subject.
    Read,
    /// Write (append) events.
    Write,
    /// List subject paths.
    Subjects,
    /// FTS / vector search.
    Search,
    /// Admin operations (mint tokens, manage peers).
    Admin,
}

impl Operation {
    /// Wire-format string used in the JSON body of `/v1/grant` and
    /// the wire-protocol `Grant` verb.
    pub fn as_wire_str(self) -> &'static str {
        match self {
            Self::Read => "read",
            Self::Write => "write",
            Self::Subjects => "subjects",
            Self::Search => "search",
            Self::Admin => "admin",
        }
    }
}

impl std::fmt::Display for Operation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_wire_str())
    }
}

/// `GET /health` response body.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct HealthInfo {
    /// Always `"ok"` on a healthy daemon.
    pub status: String,
    /// Daemon package version (semver, e.g. `"0.3.0"`).
    pub version: String,
}

/// `GET /v1/stats` response body. Future versions of the daemon may
/// add fields; the SDK accepts unknown fields silently.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct StatsInfo {
    /// Number of distinct subjects with at least one event.
    pub subject_count: i64,
}

/// `GET /v1/peers` response item.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct PeerInfo {
    /// Local identifier for this peer.
    pub peer_id: String,
    /// Address we dial when replicating with this peer.
    pub url: String,
    /// Remote peer's Ed25519 public key, hex-encoded lowercase.
    pub public_key: String,
    /// Subject globs we are willing to deliver to this peer.
    pub subject_patterns: Vec<String>,
    /// RFC3339 timestamp the peer was first registered.
    pub added_at: String,
    /// RFC3339 timestamp of the last successful exchange. Reserved
    /// for v0.4; always `None` today.
    #[serde(default)]
    pub last_seen_at: Option<String>,
}

#[derive(Serialize)]
struct GrantRequest {
    subject: String,
    operations: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    expires_in_secs: Option<i64>,
}

#[derive(Deserialize)]
struct GrantResponse {
    token: String,
}

#[derive(Deserialize)]
struct PeerListResponse {
    peers: Vec<PeerInfo>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn operation_strings_match_wire() {
        assert_eq!(Operation::Read.as_wire_str(), "read");
        assert_eq!(Operation::Write.as_wire_str(), "write");
        assert_eq!(Operation::Subjects.as_wire_str(), "subjects");
        assert_eq!(Operation::Search.as_wire_str(), "search");
        assert_eq!(Operation::Admin.as_wire_str(), "admin");
    }

    #[test]
    fn base_url_strips_trailing_slash() {
        let c = HttpAdminClient::new("http://localhost:7777/").expect("new");
        assert_eq!(c.base, "http://localhost:7777");
        let c = HttpAdminClient::new("http://localhost:7777").expect("new");
        assert_eq!(c.base, "http://localhost:7777");
    }

    #[test]
    fn with_token_attaches_token() {
        let c = HttpAdminClient::new("http://localhost:7777")
            .expect("new")
            .with_token("Y2Fw");
        assert_eq!(c.token.as_deref(), Some("Y2Fw"));
    }
}
